README.md

```diff
@@ -1,137 +1,326 @@
+)
+ament_package()
 
 ```
 
-You should see an output like this:
-`[INFO] [1656018676.015816509] [admittance_controller]: Control frame is: 'ee_link'`
+The sample program has two primary stages. The initialization stage runs only once at the beginning of the program. The
+update stage runs an arbitrary number times until the program exits.
 
-## ROS 2 CLI
-Run the following:
+## Create yaml parameter codegen file
 
-`ros2 param list`
+YAML should be a familiar language to ROS developers as it is used to specify parameter
+configurations. It also has a minimal and human-readable syntax, which make it a viable candidate for specifying code
+generation. The code generation syntax used by this package is defined in the following subsections.
 
-You should see:
+### Name space specification
 
+The generated c++ code contains a struct with type `Params` and a class of type `ParamListener`. Each are generated within specified name space. YAML allows key value pairs to be defined in a hierarchical way. The top level key will be used as the name space.
 
+```yaml
+cpp_name_space:
+# additionally fields  ...
 ```
-  /admittance_controller:
-  admittance.damping_ratio
-  admittance.mass
-  admittance.selected_axes
-  admittance.stiffness
-  chainable_command_interfaces
-  command_interfaces
-  control.frame.external
-  control.frame.id
-  enable_parameter_update_without_reactivation
-  fixed_world_frame.frame.external
-  fixed_world_frame.frame.id
-  ft_sensor.frame.external
-  ft_sensor.frame.id
-  ft_sensor.name
-  gravity_compensation.CoG.force
-  gravity_compensation.CoG.pos
-  gravity_compensation.frame.external
-  gravity_compensation.frame.id
-  joints
-  kinematics.base
-  kinematics.plugin_name
-  kinematics.tip
-  qos_overrides./parameter_events.publisher.depth
-  qos_overrides./parameter_events.publisher.durability
-  qos_overrides./parameter_events.publisher.history
-  qos_overrides./parameter_events.publisher.reliability
-  state_interfaces
-  use_sim_time
-  ```
-
-  All parameter are automatically declared and callbacks are setup by default. You can set a parameter by typing:
-
-  `ros2 param set /admittance_controller control.frame.id new_frame`
-
-  You should see:
-
-  `[INFO] [1656019001.515820371] [admittance_controller]: Control frame is: 'new_frame'`
-
-  Congratulations, you updated the parameter!
-
- If you try to set a parameter that is read only, you will get an error. Running the following
-
-  `ros2 param set /admittance_controller joints ["joint_new"]`
-
-  will result in the error
-
-  `Setting parameter failed: parameter 'joints' cannot be set because it is read-only`
 
- Running the following
+### Parameter type definition
 
- `ros2 param describe /admittance_controller admittance.damping_ratio`
+The YAML syntax can be thought of as a tree since it allows for arbitrary nesting of key value pairs. For clarity, the
+last non-nested value is referred to as a leaf. A leaf has the following required format.
```

Comment:
```suggestion
The YAML syntax can be thought of as a tree since it allows for arbitrary nesting of key-value pairs. For clarity, the last non-nested value is referred to as a leaf. 
A leaf represents a single parameter and has the following format.
```